File: scratch/gopurs-adt-reuse-validation-20260916/after-pure/src/Test/RBTreeFFICheatcode.js

class RBNodeCheat {
  constructor(c, l, v, r) {
    this.color = c;
    this.left = l;
    this.value = v;
    this.right = r;
  }
}
export const runRBTreeFFICheatcode = function(limit) {
  function balance(c, a, x, b) {
    if (c === 1) { // black
        if (a !== null && a.color === 0) {
            if (a.left !== null && a.left.color === 0) {
                return new RBNodeCheat(0, new RBNodeCheat(1, a.left.left, a.left.value, a.left.right), a.value, new RBNodeCheat(1, a.right, x, b));
            }
            if (a.right !== null && a.right.color === 0) {
                return new RBNodeCheat(0, new RBNodeCheat(1, a.left, a.value, a.right.left), a.right.value, new RBNodeCheat(1, a.right.right, x, b));
            }
        }
        if (b !== null && b.color === 0) {
            if (b.left !== null && b.left.color === 0) {
                return new RBNodeCheat(0, new RBNodeCheat(1, a, x, b.left.left), b.left.value, new RBNodeCheat(1, b.left.right, b.value, b.right));
            }
            if (b.right !== null && b.right.color === 0) {
                return new RBNodeCheat(0, new RBNodeCheat(1, a, x, b.left), b.value, new RBNodeCheat(1, b.right.left, b.right.value, b.right.right));
            }
        }
    }
    return new RBNodeCheat(c, a, x, b);
  }

  function ins(x, t) {
      if (t === null) return new RBNodeCheat(0, null, x, null);
      if (x < t.value) return balance(t.color, ins(x, t.left), t.value, t.right);
      if (x > t.value) return balance(t.color, t.left, t.value, ins(x, t.right));
      return t;
  }

  function insert(x, t) {
      let res = ins(x, t);
      return new RBNodeCheat(1, res.left, res.value, res.right);
  }

  function depth(t) {
      if (t === null) return 0;
      let ld = depth(t.left);
      let rd = depth(t.right);
      return 1 + (ld > rd ? ld : rd);
  }

  let n = Math.floor(limit);
  let acc = null;
  for (let i = n; i > 0; i--) {
      acc = insert(i, acc);
  }
  return depth(acc);
};

File: scratch/compact-guards-integration-20260916/compact/src/Test/RBTreeFFICheatcode.rs
use std::rc::Rc;

type Tree = Option<Rc<Node>>;

#[derive(Clone)]
struct Node {
    color: u8,
    left: Tree,
    value: i64,
    right: Tree,
}

fn node(color: u8, left: Tree, value: i64, right: Tree) -> Rc<Node> {
    Rc::new(Node { color, left, value, right })
}

fn balance(c: u8, a: Tree, x: i64, b: Tree) -> Rc<Node> {
    if c == 1 { // black
        if let Some(a_node) = &a {
            if a_node.color == 0 { // red
                if let Some(al) = &a_node.left {
                    if al.color == 0 {
                        let left = node(1, al.left.clone(), al.value, al.right.clone());
                        let right = node(1, a_node.right.clone(), x, b);
                        return node(0, Some(left), a_node.value, Some(right));
                    }
                }
                if let Some(ar) = &a_node.right {
                    if ar.color == 0 {
                        let left = node(1, a_node.left.clone(), a_node.value, ar.left.clone());
                        let right = node(1, ar.right.clone(), x, b);
                        return node(0, Some(left), ar.value, Some(right));
                    }
                }
            }
        }
        if let Some(b_node) = &b {
            if b_node.color == 0 {
                if let Some(bl) = &b_node.left {
                    if bl.color == 0 {
                        let left = node(1, a, x, bl.left.clone());
                        let right = node(1, bl.right.clone(), b_node.value, b_node.right.clone());
                        return node(0, Some(left), bl.value, Some(right));
                    }
                }
                if let Some(br) = &b_node.right {
                    if br.color == 0 {
                        let left = node(1, a, x, b_node.left.clone());
                        let right = node(1, br.left.clone(), br.value, br.right.clone());
                        return node(0, Some(left), b_node.value, Some(right));
                    }
                }
            }
        }
    }
    node(c, a, x, b)
}

fn ins(x: i64, t: &Tree) -> Rc<Node> {
    let Some(node) = t else {
        return self::node(0, None, x, None);
    };
    if x < node.value {
        let left = ins(x, &node.left);
        balance(node.color, Some(left), node.value, node.right.clone())
    } else if x > node.value {
        let right = ins(x, &node.right);
        balance(node.color, node.left.clone(), node.value, Some(right))
    } else {
        node.clone()
    }
}

fn depth(t: &Tree) -> i64 {
    let Some(node) = t else { return 0; };
    let l = depth(&node.left);
    let r = depth(&node.right);
    1 + std::cmp::max(l, r)
}

pub fn Test_RBTreeFFICheatcode_runRBTreeFFICheatcode(limit: i64) -> i64 {
    let mut acc = None;
    let mut i = limit;
    while i > 0 {
        let mut res = ins(i, &acc);
        Rc::make_mut(&mut res).color = 1; // root is always black
        acc = Some(res);
        i -= 1;
    }
    depth(&acc)
}
